pyiceberg/table/name_mapping.py

```diff
@@ -0,0 +1,203 @@
+        """Return the number of mappings."""
+        return len(self.root)
+
+    def __str__(self) -> str:
+        """Convert the name-mapping into a nicely formatted string."""
+        if len(self.root) == 0:
+            return "[]"
+        else:
+            return "[\n  " + "\n  ".join([str(e) for e in self.root]) + "\n]"
+
+
+T = TypeVar("T")
+
+
+class NameMappingVisitor(Generic[T], ABC):
+    @abstractmethod
+    def mapping(self, nm: NameMapping, field_results: T) -> T:
+        """Visit a NameMapping."""
+
+    @abstractmethod
+    def fields(self, struct: List[MappedField], field_results: List[T]) -> T:
+        """Visit a List[MappedField]."""
+
+    @abstractmethod
+    def field(self, field: MappedField, field_result: T) -> T:
+        """Visit a MappedField."""
+
+
+class _IndexById(NameMappingVisitor[Dict[int, MappedField]]):
+    result: Dict[int, MappedField]
+
+    def __init__(self) -> None:
+        self.result = {}
+
+    def mapping(self, nm: NameMapping, field_results: Dict[int, MappedField]) -> Dict[int, MappedField]:
+        return field_results
+
+    def fields(self, struct: List[MappedField], field_results: List[Dict[int, MappedField]]) -> Dict[int, MappedField]:
+        return self.result
+
+    def field(self, field: MappedField, field_result: Dict[int, MappedField]) -> Dict[int, MappedField]:
+        if field.field_id in self.result:
+            raise ValueError(f"Invalid mapping: ID {field.field_id} is not unique")
+
+        self.result[field.field_id] = field
+
+        return self.result
+
+
+class _IndexByName(NameMappingVisitor[Dict[str, MappedField]]):
+    def mapping(self, nm: NameMapping, field_results: Dict[str, MappedField]) -> Dict[str, MappedField]:
+        return field_results
+
+    def fields(self, struct: List[MappedField], field_results: List[Dict[str, MappedField]]) -> Dict[str, MappedField]:
+        return dict(ChainMap(*field_results))
+
+    def field(self, field: MappedField, field_result: Dict[str, MappedField]) -> Dict[str, MappedField]:
+        result: Dict[str, MappedField] = {
+            f"{field_name}.{key}": result_field for key, result_field in field_result.items() for field_name in field.names
+        }
+
+        for name in field.names:
+            result[name] = field
+
+        return result
+
+
+@singledispatch
+def visit_name_mapping(obj: Union[NameMapping, List[MappedField], MappedField], visitor: NameMappingVisitor[T]) -> T:
+    """Traverse the name mapping in post-order traversal."""
+    raise NotImplementedError(f"Cannot visit non-type: {obj}")
+
+
+@visit_name_mapping.register(NameMapping)
+def _(obj: NameMapping, visitor: NameMappingVisitor[T]) -> T:
+    return visitor.mapping(obj, visit_name_mapping(obj.root, visitor))
+
+
+@visit_name_mapping.register(list)
+def _(fields: List[MappedField], visitor: NameMappingVisitor[T]) -> T:
+    results = [visitor.field(field, visit_name_mapping(field.fields, visitor)) for field in fields]
```

Comment:
What happens if `fields` is `None`? Will that return `None` or raise an exception?

Reply:
I think this is hidden by returning `[]` instead of `None` in the `_CreateMapping` visitor.

Reply:
Yes, this is an interesting semantic discussion. When Pydantic deserializes, and it encounters that the `fields` field is missing, it will automatically inject an empty list:

https://github.com/apache/iceberg-python/blob/623ad6ad2091583bc4ea0993912dbb830ea8d72e/pyiceberg/table/name_mapping.py#L40

When serializing to JSON, it will omit the `fields` key in the JSÓN object.

This way we can just always assume that there is a list as it is annotated, and this is also why the type-checker isn't complaining. So it boils down to the definition: Is there a difference in meaning when `fields` is set to `None` or `[]`?

Reply:
I also added tests to illustrate the (de)serialization and also capture this behavior.

Reply:
As long as it works, I'm good with this.